eth/eth.go

```diff
@@ -209,53 +207,46 @@ func ValidateBlobsSidecar(slot Slot, beaconBlockRoot Root, expectedKZGCommitment
 // https://github.com/ethereum/consensus-specs/blob/dev/specs/eip4844/beacon-chain.md#tx_peek_blob_versioned_hashes
 //
 // Format of the blob tx relevant to this function is as follows:
-//   0: type (value should always be BlobTxType, 1 byte)
-//   1: message offset (value should always be 69, 4 bytes)
-//   5: ECDSA signature (65 bytes)
-//   70: start of "message" (192 bytes)
-//     258: start of the versioned hash offset within "message"  (4 bytes)
-//   262-: rest of the tx following message
+//
+//		0: type (value should always be BlobTxType)
+//		1: message offset: 4 bytes
+//		5: ECDSA signature: 65 bytes
+//		70: start of "message": 192 bytes
+//			70: chain_id: 32 bytes
+//			102: nonce: 8 bytes
+//			110: priority_fee_per_gas: 32 bytes
+//			142: max_basefee_per_gas: 32 bytes
+//			174: gas: 8 bytes
+//			182: to: 4 bytes - offset (relative to "message")
+//			186: value: 32 bytes
+//			218: data: 4 bytes - offset (relative to "message")
+//			222: access_list: 4 bytes - offset (relative to "message")
+//			226: max_fee_per_data_gas: 32 bytes
+//			258: blob_versioned_hashes: 4 bytes - offset (relative to "message")
+//	     262: start of dynamic data of "message"
+//
+// This function does not fully verify the encoding of the provided tx, but will sanity-check the tx type,
+// and will never panic on malformed inputs.
 func TxPeekBlobVersionedHashes(tx []byte) ([]VersionedHash, error) {
 	// we start our reader at the versioned hash offset within the serialized tx
-	if len(tx) < BlobVersionedHashesOffset {
+	if len(tx) < BlobVersionedHashesOffset+4 {
 		return nil, errors.New("blob tx invalid: too short")
 	}
 	if tx[0] != BlobTxType {
 		return nil, errors.New("invalid blob tx type")
 	}
-	dr := codec.NewDecodingReader(bytes.NewReader(tx[BlobVersionedHashesOffset:]), uint64(len(tx)-BlobVersionedHashesOffset))
-
-	// read the offset to the versioned hashes
-	var offset uint32
-	offset, err := dr.ReadOffset()
-	if err != nil {
-		return nil, fmt.Errorf("could not read versioned hashes offset: %v", err)
-	}
-
-	// Advance dr to the versioned hash list. We subtract blobMessageLen from the offset here to
-	// account for the fact that the offset is relative to the position of "message" (70) and we
-	// are currently positioned at the end of it (262).
-	skip := uint64(offset) - blobMessageLen
-	skipped, err := dr.Skip(skip)
-	if err != nil {
-		return nil, fmt.Errorf("could not skip to versioned hashes: %v", err)
+	offset := binary.LittleEndian.Uint32(tx[BlobVersionedHashesOffset : BlobVersionedHashesOffset+4])
+	if uint64(offset) > uint64(len(tx)) {
```

Comment:
is this right? offset is relative to tx[70] not tx[0] I believe, so perhaps you just need to add 70 to it first.

Reply:
nice catch, will fix

Reply:
updated, ready for review. I think we need some standard json tests for this, like we did with BLS: https://github.com/ethereum/bls12-381-tests/ then we can catch it before plugging it into a client (with its own testing that should cover this).

Reply:
lg, when i move the tests still in the geth fork over I can add something for this function.